hedera-node/src/main/java/com/hedera/services/txns/contract/ContractUpdateTransitionLogic.java

```diff
@@ -21,77 +21,122 @@
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INVALID_ADMIN_KEY;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INVALID_EXPIRATION_TIME;
 import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INVALID_RENEWAL_PERIOD;
 import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.OK;
-import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.SUCCESS;
 
 public class ContractUpdateTransitionLogic implements TransitionLogic {
-	private static final Logger log = LogManager.getLogger(ContractUpdateTransitionLogic.class);
-
-	private final HederaLedger ledger;
-	private final OptionValidator validator;
 	private final TransactionContext txnCtx;
-	private final UpdateCustomizerFactory customizerFactory;
-	private final Supplier<MerkleMap<EntityNum, MerkleAccount>> contracts;
+	private final OptionValidator validator;
+	private final AccountStore accountStore;
 
 	private final Function<TransactionBody, ResponseCodeEnum> SEMANTIC_CHECK = this::validate;
 
 	public ContractUpdateTransitionLogic(
-			HederaLedger ledger,
-			OptionValidator validator,
 			TransactionContext txnCtx,
-			UpdateCustomizerFactory customizerFactory,
-			Supplier<MerkleMap<EntityNum, MerkleAccount>> contracts
+			OptionValidator validator,
+			AccountStore accountStore
 	) {
-		this.ledger = ledger;
-		this.validator = validator;
 		this.txnCtx = txnCtx;
-		this.contracts = contracts;
-		this.customizerFactory = customizerFactory;
+		this.validator = validator;
+		this.accountStore = accountStore;
 	}
 
 	@Override
 	public void doStateTransition() {
-		try {
-			var contractUpdateTxn = txnCtx.accessor().getTxn();
-			var op = contractUpdateTxn.getContractUpdateInstance();
-			var id = op.getContractID();
-			var target = contracts.get().get(fromContractId(id));
-
-			var result = customizerFactory.customizerFor(target, validator, op);
-			var customizer = result.getLeft();
-			if (customizer.isPresent()) {
-				ledger.customize(asAccount(id), customizer.get());
-				txnCtx.setStatus(SUCCESS);
-			} else {
-				txnCtx.setStatus(result.getRight());
-			}
-		} catch (Exception e) {
-			log.warn("Avoidable exception!", e);
-			txnCtx.setStatus(FAIL_INVALID);
+		/* --- Extract gRPC --- */
+		final var contractUpdateTxn = txnCtx.accessor().getTxn();
+		final var op = contractUpdateTxn.getContractUpdateInstance();
+		final var targetId = Id.fromGrpcContract(op.getContractID());
+
+		/* --- Translate the account --- */
+		final var target = accountStore.loadContract(targetId);
+		final var candidate = op.getAdminKey();
+
+		/* --- Conditions --- */
+		final var isSentinel = candidate.equals(IMMUTABILITY_SENTINEL_KEY);
+
+		/* --- Validate --- */
+		if (op.hasAdminKey()) {
+			validateFalse(isValidAdminKey(candidate, isSentinel), INVALID_ADMIN_KEY);
+		}
+		if (op.hasExpirationTime()) {
+			validateFalse(!validator.isValidExpiry(op.getExpirationTime()), INVALID_EXPIRATION_TIME);
```

Comment:
validating the negation of a negation seems confusing.
`validateTrue(validator.isValidExpiry(op.getExpirationTime()), INVALID_EXPIRATION_TIME);` might be better in this case.

Reply:
solved